### PR TITLE
feat: add CUDA kernels that calculate length/sum

### PR DESCRIPTION
**Pull request by ManasviGoyal**
- New kernels added -
  1.  awkward_Index_nones_as_index
  2. awkward_ByteMaskedArray_numnull
  3. awkward_IndexedArray_numnull
  4. awkward_IndexedArray_numnull_parents
  5. awkward_ListArray_getitem_jagged_carrylen
  6. awkward_ListArray_getitem_next_range_counts
  7. awkward_ListArray_rpad_and_clip_length_axis1
  8. awkward_ListOffsetArray_reduce_nonlocal_nextstarts_64
  9. awkward_RegularArray_getitem_next_array_regularize
  10. awkward_RegularArray_reduce_local_nextparents
  11. awkward_RegularArray_reduce_nonlocal_preparenext
  12. awkward_sorting_ranges_length

- Makes `awkward_ListArray_min_range` more efficient by using `cupy.min` #2987 
- Fixes the implemetation of many existing kernels
- Adds unit-tests

### CODE REVIEW

**Comment by ManasviGoyal** [comment]
@jpivarski I am done with this PR but I am not quite sure why the integration test is failing. All the tests pass locally.

Edit: Oh, I just noticed #2998 which refers to this issue. I need to wait for that PR to merge.

File: src/awkward/_connect/cuda/cuda_kernels/awkward_IndexedArray_reduce_next_nonlocal_nextshifts_64.cu
Comment (jpivarski) [inline src/awkward/_connect/cuda/cuda_kernels/awkward_IndexedArray_reduce_next_nonlocal_nextshifts_64.cu]: This originally had black/ruff-style formatting so that changes in diff would be more localized. For instance, a change in the name would only affect the line where the name is, but with this indenting, a change in the length of the name would change where all arguments are indented to.

This is not a major issue. It's very unlikely that we'll be changing names or argument lists, since it's all constrained by a specification and users don't see it. I'm just saying that there was a reason for the original formatting.

**Review by jpivarski — APPROVED** [review]
These look great! I also tested everything on my GPU; there were no errors. We should _not merge_ until after 2.6.0, though.

File: src/awkward/_connect/cuda/cuda_kernels/awkward_IndexedArray_reduce_next_nonlocal_nextshifts_64.cu
Comment (ManasviGoyal) [reply src/awkward/_connect/cuda/cuda_kernels/awkward_IndexedArray_reduce_next_nonlocal_nextshifts_64.cu]: Oh, since all the other kernels had the other formatting except these few so I changed these too. But I can fix that in another PR (or maybe in this one only?). It would be easier for me too to use this formatting for new kernels.

**Comment by ManasviGoyal** [comment]
> These look great! I also tested everything on my GPU; there were no errors. We should _not merge_ until after 2.6.0, though.

@jpivarski Can this be merged now?

**Comment by jpivarski** [comment]
Thanks for formatting it. In the changes since I last viewed, it looks like a lot more kernels were touched. Was it the case that only a few of them were previously formatted in black-style, and now you did all of them?

I think this PR will change `awkward-cpp`, and I want to release another `awkward` version, 2.6.1 with #3006, so please hold off on merging this for a little while longer. Thanks!

**Comment by ManasviGoyal** [comment]
> Was it the case that only a few of them were previously formatted in black-style, and now you did all of them?

Yes, there were very few that were formatted earlier.

> I think this PR will change `awkward-cpp`, and I want to release another `awkward` version, 2.6.1 with #3006, so please hold off on merging this for a little while longer. Thanks!

Sure. Thank you.

**Comment by jpivarski** [comment]
[Awkward 2.6.1](https://github.com/scikit-hep/awkward/releases/tag/v2.6.1) is out, so this can be merged whenever you want. It is understood that the next version will require a new awkward-cpp. (Unfortunately, that might slow down testing of #3007. Oh well.)

**Comment by ManasviGoyal** [comment]
@jpivarski `pre-commit` is failing after merging with `main`. Should I wait for it to be fixed before merging?

**Comment by agoose77** [comment]
@ManasviGoyal if you can hold off on merging, I'll fix these in a separate PR and merge them into main :) Will be about an hour all-in.

**Comment by ManasviGoyal** [comment]
> @ManasviGoyal if you can hold off on merging, I'll fix these in a separate PR and merge them into main :) Will be about an hour all-in.

Sure. Thanks!

**Pull request merged**
